aws-s3-transfer-manager/tests/upload_checksum_test.rs

```diff
@@ -0,0 +1,829 @@
+    operation::upload::{ChecksumStrategy, UploadOutput},
+    types::{ConcurrencySetting, PartSize},
+};
+use aws_sdk_s3::{
+    operation::{
+        complete_multipart_upload::CompleteMultipartUploadOutput,
+        create_multipart_upload::CreateMultipartUploadOutput, put_object::PutObjectOutput,
+        upload_part::UploadPartOutput,
+    },
+    types::{ChecksumAlgorithm, ChecksumType},
+};
+use aws_smithy_mocks_experimental::{mock, Rule, RuleMode};
+use aws_smithy_runtime::test_util::capture_test_logs::capture_test_logs;
+use bytes::Bytes;
+use pin_project_lite::pin_project;
+use test_common::mock_client_with_stubbed_http_client;
+
+const MEBIBYTE: usize = 1024 * 1024;
+const PART_SIZE: usize = 5 * MEBIBYTE;
+
+fn calculate_checksum(algorithm: &ChecksumAlgorithm, data: &[u8]) -> String {
+    let smithy_algorithm =
+        aws_smithy_checksums::ChecksumAlgorithm::from_str(algorithm.as_str()).unwrap();
+    let mut http_checksum = smithy_algorithm.into_impl();
+    http_checksum.update(data);
+    http_checksum.header_value().to_str().unwrap().to_string()
+}
+
+fn calculate_full_object_checksum(algorithm: &ChecksumAlgorithm, parts: &Vec<Bytes>) -> String {
+    // S3 doesn't allow SHA algorithms to do full-object checksums with multipart upload
+    assert_ne!(algorithm, &ChecksumAlgorithm::Sha1);
+    assert_ne!(algorithm, &ChecksumAlgorithm::Sha256);
+
+    let smithy_algorithm =
+        aws_smithy_checksums::ChecksumAlgorithm::from_str(algorithm.as_str()).unwrap();
+    let mut http_checksum = smithy_algorithm.into_impl();
+    for part in parts {
+        http_checksum.update(part.as_ref());
+    }
+    http_checksum.header_value().to_str().unwrap().to_string()
+}
+
+fn calculate_composite_checksum(algorithm: &ChecksumAlgorithm, parts: &Vec<Bytes>) -> String {
+    // S3 doesn't allow CRC64NVME to do composite checksums
+    assert_ne!(algorithm, &ChecksumAlgorithm::Crc64Nvme);
+
+    let mut concatenated_part_checksums = String::new();
+    for part in parts {
+        let checksum = calculate_checksum(algorithm, part.as_ref());
+        concatenated_part_checksums.push_str(&checksum);
+    }
+
+    let composite_checksum = calculate_checksum(algorithm, concatenated_part_checksums.as_bytes());
+    let count = parts.len();
+    format!("{composite_checksum}-{count}")
+}
+
+/// Calculate final Composite or FullObject checksum for multipart upload
+fn calculate_multipart_checksum(
+    algorithm: &ChecksumAlgorithm,
+    checksum_type: &ChecksumType,
+    parts: &Vec<Bytes>,
+) -> String {
+    match checksum_type {
+        ChecksumType::Composite => calculate_composite_checksum(algorithm, parts),
+        ChecksumType::FullObject => calculate_full_object_checksum(algorithm, parts),
+        _ => panic!("unrecognized checksum type"),
+    }
+}
+
+/// Calculate ETag as it appears in S3 headers, with the extra quotes and everything
+fn calculate_etag(data: &[u8]) -> String {
+    let digest = md5::compute(data);
```

Comment:
Am I missing something or is `md5` crate missing from our dependencies in `Cargo.toml`?

Reply:
Yeah, my Cargo.toml has a bunch of edits to use a local version of the SDK so I haven't committed it